src/mapstorage/XmlMapStorage.cpp

```diff
@@ -41,20 +214,396 @@ void XmlMapStorage::newData()
+                << name << "> inside <exit>";
+        }
+        skipXmlElement(stream);
+    }
+    exit.setDoorFlags(doorFlags);
+    // EXIT flag is almost always set, thus we save it inverted.
+    exit.setExitFlags(exitFlags ^ ExitFlagEnum::EXIT);
+}
+
+// check that all rooms' exits "to" actually point to an existing room,
+// and add matching exits "from"
+void XmlMapStorage::connectRoomsExitFrom(QXmlStreamReader &stream)
+{
+    for (auto &elem : loadedRooms) {
+        Room &room = deref(elem.second);
+        for (const ExitDirEnum dir : ALL_EXITS7) {
+            connectRoomExitFrom(stream, room, dir);
+        }
+    }
+}
+
+void XmlMapStorage::connectRoomExitFrom(QXmlStreamReader &stream,
+                                        const Room &fromRoom,
+                                        ExitDirEnum dir)
+{
+    const Exit &fromE = fromRoom.exit(dir);
+    if (fromE.outIsEmpty()) {
+        return;
+    }
+    const RoomId fromId = fromRoom.getId();
+    for (const RoomId toId : fromE.outRange()) {
+        auto iter = loadedRooms.find(toId);
+        if (iter == loadedRooms.end()) {
+            throwErrorFmt(stream,
+                          "room %1 has exit %2 to non-existing room %3",
+                          roomIdToString(fromId),
+                          lowercaseDirection(dir),
+                          roomIdToString(toId));
+        }
+        Room &toRoom = deref(iter->second);
+        toRoom.addInExit(opposite(dir), fromId);
+    }
+}
+
+// add all loaded rooms to m_mapData
+void XmlMapStorage::moveRoomsToMapData()
+{
+    for (auto &elem : loadedRooms) {
+        m_mapData.insertPredefinedRoom(elem.second);
+    }
+    loadedRooms.clear();
+}
+
+void XmlMapStorage::loadMarker(QXmlStreamReader &stream)
+{
+    const QXmlStreamAttributes attrs = stream.attributes();
+    bool fail = false;
+    const InfoMarkTypeEnum type = conv.toEnum<InfoMarkTypeEnum>(attrs.value("type"), fail);
+    const InfoMarkClassEnum clas = conv.toEnum<InfoMarkClassEnum>(attrs.value("class"), fail);
+    if (fail) {
+        throwErrorFmt(stream,
+                      "invalid marker attributes type=\"%1\" class=\"%2\"",
+                      attrs.value("type").toString(),
+                      attrs.value("class").toString());
+    }
+    QStringRef anglestr = attrs.value("angle");
+    int angle = 0;
+    if (!anglestr.isNull()) {
+        angle = conv.toNumber<int>(anglestr, fail);
+        if (fail) {
+            throwErrorFmt(stream, "invalid marker attribute angle=\"%1\"", anglestr.toString());
+        }
+    }
+
+    SharedInfoMark sharedmarker = InfoMark::alloc(m_mapData);
+    InfoMark &marker = deref(sharedmarker);
+
+    marker.setType(type);
+    marker.setClass(clas);
+    marker.setRotationAngle(angle);
+
+    while (stream.readNextStartElement() && !stream.hasError()) {
+        const QStringRef name = stream.name();
+        if (name == "pos1") {
+            marker.setPosition1(loadCoordinate(stream));
```

Comment:
What's the sanest default for mark type and mark class?
We should use that.

A mark without position currently defaults to 0,0,0 which is admittedly less useful than rooms defaulting at 0,0,0.
Still, it seems better to me than rejecting the whole map because of a missing coordinate.

Reply:
I think we should reject the map. The power user should edit it to fix the error rather than it silently corrupting more over time. I don't see this XML mode being something most of our normal users use so we should be more strict given the risk it brings.

Reply:
Done. I wrote the comment "marker pos1 is now mandatory" in the wrong place (below).
Marker type and class are mandatory too.

Also, as it happens quite often that pos1 and pos2 are equal, I updated `saveMarker` to omit pos2 if it's equal to pos1.
And correspondigly now marker pos2 is optional in `loadMarker`, and defaults to pos1.